include/pybind11/cast.h

```diff
@@ -869,7 +871,7 @@ template <typename type> class type_caster_base : public type_caster_generic {
 
     static handle cast(const itype &src, return_value_policy policy, handle parent) {
         if (policy == return_value_policy::automatic || policy == return_value_policy::automatic_reference)
-            policy = return_value_policy::copy;
+            policy = return_value_policy::reference;
```

Comment:
This code path occurs when:
(1) the return type is a const/non-const lvalue reference and no rvp is specified.
(2) a reference value is provided as a default for `py::arg`.
(3) any others?

The previous `rvp::copy` policy was ignored due to the early-return logic. With the new conditional guard, we now use `rvp::reference` and end up checking the value.

Case (1) should be `rvp::reference`. Case(2) is handled in 6745c6a981e761d95ddbcd772c720461092f918e.